### ISSUE TITLE
Categorize Students and then Group Students According to Categories

### ISSUE THREAD

**Issue by gkapfham** [GatorGrouper/gatorgrouper#13]
Currently, the program can only assign students to a group based on a random number generator. However, it would also be useful to assign students to a group while taking into account a student's membership in a specified category. Once these categories have been provided by input, it would be possible to then automatically assign students to groups while ensuring that, whenever possible, students in the same category are not in the same group.

**Comment by delgrecoj** [comment]
We'll assume that the input is in CSV format, with an identifier for the student, followed by a series of boolean values (zeros or ones) representing membership in different groups, so each record in that file would look like `delgrecoj,0,1,0,1,1`.

We can then break the task into individual concerns:
- parsing the CSV file into some useful format (ideally something like a HashMap, that would retain both the value as well as the category it was associated with, to avoid hardcoding the ordering)
- spreading the categories as evenly as possible across the groups

I'm thinking that maybe iterating through the groups and randomly pulling one of each category might work, but that's just an initial thought.  We of course get into complicated waters when we consider that any student could be part of several categories, and that's still working with booleans alone and ignoring any sort of rating scheme (i.e. input data in the form of 'rate between 1 and 5 your comfort level with Python programming' and the like).

**Comment by delgrecoj** [comment]
Another possible approach might be a sudoku solver approach - this worked at my internship to predict which elements of two sets were not in both, to some degree of success.  Given the multiple categories issue it would be harder to implement, but certainly seems along the same lines -- we don't want multiple people in the same category in the same group, just as in sudoku you don't want multiples of the same number in the same line / row / block.

**Comment by delgrecoj** [comment]
I've gone and created a [branch](https://github.com/gkapfham/gatorgrouper/tree/feature/i-13-categories) for us to work with.

@sporrera, would you mind enhancing the program to handle CSV parsing?  It looks like a list of [dictionaries](https://docs.python.org/3/library/stdtypes.html#dict) would be a good choice of output.

@ZachAndrews98, would you mind thinking about some way to "score" groups, where a high score would represent a group with few students in the same category, and a low score would present a group with many in the same category?  There might be some nontrivial difficulties with scoring when considering students that are members of more than one category, but just a basic implementation would be good to start with, and we can expand on it later if needed.

I'm going to start thinking about different ways to generate the groups.  If we have a scoring system then we should be able to tell whether our solution is actually any better than the current randomized one.

**Comment by sporrera** [comment]
Yes- I can begin to work on that.

**Comment by sporrera** [comment]
`""" GatorGrouper randomly assigns a list of students to groups """
from random import shuffle
import argparse
import itertools
import sys
import xlrd



def read_student_file ():
	""" Reads the student identifies from the specific file,
	    returning the identifiers in a list """

	workbook = xlrd.open_workbook('categories.xlsx')
	sheet = workbook.sheet_by_name('Form Responses 1')
	data = [[sheet.cell_value(r, c) for c in range(1, sheet.ncols)] for r in range(1, sheet.nrows)]
	# Profit !

	y = ''.join([str(item) for item in data])
	print()
	lines = y.split(']')

	for i in range (0, len(lines)):
		newline = lines[i].split(', ')
		newline2 = list(newline)
	for x in range (0, len(newline)):
		if newline[x] == '\'Yes\'':
			newline2[x] = True
		elif newline[x] == '\'No\'':
			newline2[x] = False
		print(newline2[x])`

I have been working on the read_student_file.py.  So to get the categorization started, I first made a Google Form that asked students their preference of position (just as a way to test the code) which then exported into Google Sheets. I have sent both of these to you via your allegheny email addresses.

In order to import the answers of the students into read_student_file.py, I had to download and save the Google Sheet as an excel file because I have no familiarity with Google API. From there, I created a program within read_student_file.py that will print out everyone's last name followed by True and/or False boolean values corresponding to their answers, where True = Yes and False = No.

In a separate file when I practice running and implementing this code, it works fine.  Here is a picture of the output that I get where you can see the last name of the student and the boolean True/False values: 

![correct_output](https://user-images.githubusercontent.com/22895356/31577987-699faeae-b0e6-11e7-8b83-20b870112ce2.png)


However, I am having an issue returning the manipulated data in read_student_file.py. My program will print out the appropriate data but will not return it as a list. I included the error message that I am getting.

![error_message](https://user-images.githubusercontent.com/22895356/31577984-5f6a991c-b0e6-11e7-9eb7-98779f304360.png)


Does anyone have any suggestions on how to fix this or would be able to meet at some point today?

**Comment by delgrecoj** [comment]
Another team is working on a function that will pull data from the form sheet and save that data in CSV format to disk... as far as we're concerned, our function has to read that CSV and return the list of lists, which would look something like the following:

```python
[
  ["delgrecoj", true, true, true, false, true],
  ["delgrecoj", true, true, true, false, true],
  ["delgrecoj", true, true, true, false, true]
]
```

So my understanding is, that first part dealing with workbooks will not be needed, rather something that would read the CSV file that's been provided.

**Comment by delgrecoj** [comment]
I believe they said they would handle the `yes` to `true` and `no` to `false` conversions, and trimming the emails down to just the usernames.

**Comment by delgrecoj** [comment]
So our real goal seems to be something to take the following CSV file
```text
delgrecoj, true, true, true, false, true
delgrecoj, true, true, true, false, true
delgrecoj, true, true, true, false, true
```
and read it in, returning a list of lists where each list is a line of the file, like

```python
[
  ["delgrecoj", true, true, true, false, true],
  ["delgrecoj", true, true, true, false, true],
  ["delgrecoj", true, true, true, false, true]
]
```

**Comment by delgrecoj** [comment]
@dillam is your spreadsheet function going to drop 0/1 values into the CSV file, or true/false strings, or perhaps something else?

**Comment by ZachAndrews98** [comment]
@delgrecoj can you double check your changes to read_student_file, I've been getting errors from that file:
line 16, in read_student_file                                                                                              
temp[0] = record[0] # handle identifiers separately                                                                                                                                       IndexError: list assignment index out of range

**Comment by delgrecoj** [comment]
fixed.
hazards of working with Ruby, Rust, Golang, JavaScript, and Python all in the same day... all except Python are fine with that syntax...

**Comment by delgrecoj** [comment]
Alright, have a working round-robin method pushed to our feature branch, with the sudoku one to come later this evening.  I would greatly appreciate some feedback and testing from you both, @sporrera, @ZachAndrews98.  I also recently pushed a fix to the random function, which I had to rewrite to handle group sizes greater than two properly, which was dropping students in some cases of unusual group sizes.

**Comment by delgrecoj** [comment]
Also, @ZachAndrews98, if you have any thoughts on the scoring function, I would like to get that merged into our feature branch sometime tonight, to hopefully get a pull request for our entire feature branch into master setup late this evening.

**Comment by ZachAndrews98** [comment]
Round robin looks good to me, just Display prints out the students in the groups like ['gkapfham4b'] which should probably be changed. Also made the changes @delgrecoj suggested to the group_scoring file so take a look at that and let me know if you want me to change anything.

**Comment by sporrera** [comment]
@delgrecoj just making sure I still run the program with `python3 gatorgrouper.py` ? Or are there missing arguments I need to enter as well?

**Comment by ZachAndrews98** [comment]
@sporrera to run the round robin its python3 gatorgrouper.py --round-robin

**Comment by sporrera** [comment]
Thanks for the input @ZachAndrews98 . I got it working and the output displays correctly @delgrecoj

**Comment by delgrecoj** [comment]
On my systems it's just `python` and `python3` doesn't exist, but at any rate, thanks for taking a look.

Does that method seem alright for dealing with the potentially smaller-than-specified last group?  I had to do something since @gkapfham only had dealt with the case of group sizes of two and one student leftover, and it seemed most natural to just distribute them through the rest of the groups evenly.

**Comment by delgrecoj** [comment]
It might raise some eyebrows when people ask for a group size of say four and are given mostly 5-person groups or something similar... could perhaps add some threshold to distribute them that would allow for a mostly full partial group but not for a mostly empty partial group, but not sure what the requirements are for this... probably something to meet with the "customer" about.

**Comment by ZachAndrews98** [comment]
Yeah, I think allowing there to be one group that is slightly smaller than the rest should be fine, but it definitely is a good idea to check with the customer.

**Comment by delgrecoj** [comment]
Merged the scoring changes into #34, and created a new issue #40 for feedback on how we should move forward.

**Comment by delgrecoj** [comment]
merged feature branch into master as of pull request #41.
